### PR TITLE
Integrate InheritDoc into the build

### PR DESCRIPTION
**Pull request by christothes**
# Summary
Adds a reference to [InheritDoc](https://www.nuget.org/packages/SauceControl.InheritDoc/) to replace all `/// <inheritdoc/>` comments with the actual comment content in the documentation xml produced by the build.

This is a development-only dependency; it will not be deployed with or referenced by the libraries.

Note this produces a lot of warnings for `<inheritdoc />` comments that need to be further refined with a `cref="somBaseType"` attribute. Some examples of this are included in this PR.

Example for `Azure.Security.KeyVault.Administration.xml`:

**Before**
```xml
<member name="P:Azure.Security.KeyVault.Administration.RestoreOperation.Id">
    <inheritdoc/>
</member>
<member name="P:Azure.Security.KeyVault.Administration.RestoreOperation.Value">
    <inheritdoc/>
</member>
<member name="P:Azure.Security.KeyVault.Administration.RestoreOperation.HasCompleted">
    <inheritdoc/>
</member>
<member name="P:Azure.Security.KeyVault.Administration.RestoreOperation.HasValue">
    <inheritdoc/>
</member>
<member name="M:Azure.Security.KeyVault.Administration.RestoreOperation.GetRawResponse">
    <inheritdoc/>
</member>
```

**After**
```xml
<member name="P:Azure.Security.KeyVault.Administration.RestoreOperation.Id">
    <summary>
    Gets an ID representing the operation that can be used to poll for
    the status of the long-running operation.
    </summary>
</member>
<member name="P:Azure.Security.KeyVault.Administration.RestoreOperation.Value">
    <summary>
    Final result of the long-running operation.
    </summary><remarks>
    This property can be accessed only after the operation completes successfully (HasValue is true).
    </remarks>
</member>
<member name="P:Azure.Security.KeyVault.Administration.RestoreOperation.HasCompleted">
    <summary>
    Returns true if the long-running operation completed.
    </summary>
</member>
<member name="P:Azure.Security.KeyVault.Administration.RestoreOperation.HasValue">
    <summary>
    Returns true if the long-running operation completed successfully and has produced final result (accessible by Value property).
    </summary>
</member>
<member name="M:Azure.Security.KeyVault.Administration.RestoreOperation.GetRawResponse">
    <summary>
    The last HTTP response received from the server.
    </summary><remarks>
    The last response returned from the server during the lifecycle of this instance.
    An instance of Operation<typeparamref name="T" /> sends requests to a server in UpdateStatusAsync, UpdateStatus, and other methods.
    Responses from these requests can be accessed using GetRawResponse.
    </remarks>
</member>
```

### CODE REVIEW

File: eng/Directory.Build.Data.props [hunk 1]
```diff
@@ -68,6 +69,12 @@
     </NoWarn>
   </PropertyGroup>
 
+  <PropertyGroup Condition="'$(InheritDocEnabled)' == 'true' and '$(TargetFramework)'=='netstandard2.0'">
+    <NoWarn>
+      $(NoWarn);IDT001<!-- InheritDoc related to malformed XML in netstandard.xml -->
+      </NoWarn>
```
Comment (weshaggard) [inline eng/Directory.Build.Data.props:75]:
```suggestion
    </NoWarn>
```

File: eng/Directory.Build.Data.targets [hunk 2]
```diff
@@ -222,4 +222,15 @@
   </Target>
 
   <Import Project="$(CentralPackageVersionPackagePath)\Sdk.targets" />
+
+  <!-- InheritDoc-->
+  <ItemGroup Condition="'$(InheritDocEnabled)' == 'true'">
+    <PackageReference Include="SauceControl.InheritDoc" Version="1.1.1" PrivateAssets="all" />
```
Comment (weshaggard) [inline eng/Directory.Build.Data.targets:228]: This looks to be a third party library, is it safe for use here? 

Also you will need to move the version number to the Packages.Data.props file. The build should actually be failing for this.

**Comment by pakrym** [comment]
I presume the manual comment changes are part of testing. Other than that LGTM.

File: eng/Directory.Build.Data.targets (hunk 2 above)
Comment (christothes) [reply eng/Directory.Build.Data.targets:228]: > This looks to be a third party library, is it safe for use here?

It's MIT license. Any concerns other than it is 3rd party?

Comment (weshaggard) [reply eng/Directory.Build.Data.targets]: No particular concern I just wanted to ensure we understand what dependencies we are taking. This is just a tool and not a runtime dependency so it is less of an issue but in general we have to assume 3rd party dependencies may have issues such as no maintainer in the future.

**Review by weshaggard — APPROVED** [review]

**Review by chamons — APPROVED** [review]
:+1: assuming we're fine taking on a 3rd party build dependency. 

It doesn't look too huge though, and we can drop it once docs supports inheritdoc.

Did you notice a build speed hit from this?

**Comment by christothes** [comment]
> Did you notice a build speed hit from this?

Nothing significant. If, for some reason, it ever becomes an  issue, it can be conditioanlly disabled like this:

`dotnet build /p:EnableInheritDoc="false"`

File: eng/Directory.Build.Data.props [hunk 3]
```diff
@@ -52,6 +52,7 @@
     <UpdateSourceOnBuild Condition="'$(UpdateSourceOnBuild)' == ''">$(AZURE_DEV_UPDATESOURCESONBUILD)</UpdateSourceOnBuild>
     <PowerShellExe Condition="'$(PowerShellExe)' == ''">pwsh</PowerShellExe>
     <CoverletOutputFormat Condition="'$(CoverletOutputFormat)' == '' and '$(CollectCoverage)' == 'true'">cobertura</CoverletOutputFormat>
+    <InheritDocEnabled Condition="'$(EnableInheritDoc)' == '' or '$(EnableInheritDoc)' == 'true'">true</InheritDocEnabled>
```
Comment (heaths) [inline eng/Directory.Build.Data.props:55]: Maybe just `'$(EnableInheritDoc)' != 'false'`? Also, why not just use a single variable for whether it's enabled or not. Seems a lot less to manage overtime (and easier to diagnose why something isn't working as expected.

File: eng/Directory.Build.Data.targets [hunk 4]
```diff
@@ -200,6 +200,18 @@
...
   </Target>
 
+  <!-- InheritDoc-->
+  <ItemGroup Condition="'$(InheritDocEnabled)' == 'true'">
+    <PackageReference Include="SauceControl.InheritDoc" PrivateAssets="all" />
+  </ItemGroup>
+
+  <!-- Fixup netstandard2.malformed docs issue https://github.com/saucecontrol/InheritDoc#bad-netstandard-docs -->
+  <ItemGroup Condition="'$(InheritDocEnabled)' == 'true' and '$(TargetFramework)'=='netstandard2.0'">
+    <PackageDownload Include="NETStandard.Library.Ref" Version="[2.1.0]" />
```
Comment (heaths) [inline eng/Directory.Build.Data.targets:210]: Might be a good idea to still manage this version range in Packages.Data.props so everything is nice and consolidated.

File: eng/Directory.Build.Data.targets [hunk 5]
```diff
@@ -200,6 +200,18 @@
...
+    <InheritDocReference Include="$([MSBuild]::EnsureTrailingSlash('$(NugetPackageRoot)'))netstandard.library.ref\2.1.0\ref\netstandard2.1\netstandard.xml" />
```
Comment (heaths) [inline eng/Directory.Build.Data.targets:211]: Tip: You can just include a trailing backslash. Unless the app breaks with consecutive backslashes/slashes (99% of them won't) it doesn't hurt. This is common in MSBuild and not as verbose.

File: eng/Packages.Data.props [hunk 6]
```diff
@@ -196,4 +196,8 @@
     <PackageReference Update="System.Reflection.Emit" Version="4.7.0" />
   </ItemGroup>
 
+  <!-- InheritDoc-->
+  <ItemGroup Condition="'$(InheritDocEnabled)' == 'true'">
```
Comment (heaths) [inline eng/Packages.Data.props:200]: The condition isn't necessary. If the package wasn't included the `Update` does nothing. I wouldn't duplicate conditions more than necessary as it gets harder to maintain.

**Review by heaths — APPROVED** [review]

File: eng/Directory.Build.Data.props (hunk 3 above)
Comment (christothes) [reply eng/Directory.Build.Data.props:55]: Your suggestion removed the reason I did two variables. I didn't want to duplicate the `or` logic everywhere.

File: eng/Directory.Build.Data.props [hunk 7]
```diff
@@ -52,7 +52,7 @@
     <UpdateSourceOnBuild Condition="'$(UpdateSourceOnBuild)' == ''">$(AZURE_DEV_UPDATESOURCESONBUILD)</UpdateSourceOnBuild>
     <PowerShellExe Condition="'$(PowerShellExe)' == ''">pwsh</PowerShellExe>
     <CoverletOutputFormat Condition="'$(CoverletOutputFormat)' == '' and '$(CollectCoverage)' == 'true'">cobertura</CoverletOutputFormat>
-    <InheritDocEnabled Condition="'$(EnableInheritDoc)' == '' or '$(EnableInheritDoc)' == 'true'">true</InheritDocEnabled>
+    <InheritDocEnabled Condition="'$(EnableInheritDoc)' != 'false'">true</InheritDocEnabled>
```
Comment (heaths) [inline eng/Directory.Build.Data.props:55]: You don't need this property anymore then, do you?

Comment (christothes) [reply eng/Directory.Build.Data.props:55]: I guess not. If I set it to true explicitly, would it still respect an msbuild argument to override it to false?

**Pull request merged**